clients/src/main/java/org/apache/kafka/common/utils/Time.java

```diff
@@ -86,4 +89,30 @@ default Timer timer(Duration timeout) {
         return timer(timeout.toMillis());
     }
 
+    /**
+     * Wait for a future to complete, or time out.
+     *
+     * @param future        The future to wait for.
+     * @param deadlineNs    The time in the future, in monotonic nanoseconds, to time out.
+     * @return              The result of the future.
+     * @param <T>           The type of the future.
+     */
+    default <T> T waitForFuture(
+        CompletableFuture<T> future,
+        long deadlineNs
+    ) throws TimeoutException, InterruptedException, ExecutionException  {
+        TimeoutException timeoutException = null;
+        while (true) {
```

Comment:
Why is the loop needed? Does `future.get` not respect the timeout?

Reply:
It's a bit unclear. It's traditional for condition variables to allow spurious wakeups, and probably this is implemented with a condition variable. The JavaDoc for `get` just says that it will wait "at most the given time" which doesn't really clarify matters. In any case, the loop certainly does no harm.